### PR TITLE
Add management of delivery availability in Federation settings

### PR DESCRIPTION
**Pull request by noellabo**
Add the ability to stop and resume delivery to "Federation Settings".

managements_delivery1

This is a function that makes delivery availability manageable.

- UI to check delivery error status
- Manually stop delivery
- Manually resume delivery
- Manually delete the error history

This allows you to clear a large number of queued retry jobs or mark closed servers for non-delivery without deleting them. It can be operated on the UI, allowing multiple administrators to address the issue without having direct access to the command line. This should also be useful for users of hosting services.

### CODE REVIEW

**Comment by noellabo** [comment]
Later, I'll fix the missing materialized view and add a description of the PR.

**Comment by shleeable** [comment]
yesssssss

**Comment by ClearlyClaire** [comment]
I somehow missed that PR so far. I'll try to actually review it by the end of the week, but here are a few remarks off the top of my head:
- I think it's a welcome feature, both for diagnostic purposes and managing edge cases from the admin's perspective
- I'm not too fond of the filter shortcut menu, especially on the index page
- “remove delivery errors” feels a bit weird to me… maybe “clear delivery errors” or “clear delivery errors log”?
- otherwise I think the UI is pretty nice

**Comment by noellabo** [comment]
Changed "Remove delivery errors" to "Clear delivery errors".

The default filter state is limited, which requires two filter changes to switch to the delivery warning (switch from limited to all, then to warning). Also, if you want to list them while dealing with delivery warnings, you need 3 transitions. I wish there was another good way.

File: app/lib/delivery_failure_tracker.rb [hunk 1]
```diff
@@ -44,6 +52,23 @@ def available?(url)
...
+
+    def warning_domains_map
+      warning_domains.each_with_object({}) { |domain, hash| hash[domain] = Redis.current.scard(exhausted_deliveries_key(domain)) }
+    end
+
+    private
+
+    def exhausted_deliveries_key(host)
+      "exhausted_deliveries:#{host}"
+    end
```
Comment (ClearlyClaire) [inline app/lib/delivery_failure_tracker.rb:71]: Hmmm I'm unconfortable with the class method having the same name as the instance method… does that even work?

File: app/models/instance.rb [hunk 2]
```diff
@@ -10,10 +10,13 @@
 class Instance < ApplicationRecord
   self.primary_key = :domain
 
+  attr_accessor :failure_days
+
   has_many :accounts, foreign_key: :domain, primary_key: :domain
 
   belongs_to :domain_block, foreign_key: :domain, primary_key: :domain
   belongs_to :domain_allow, foreign_key: :domain, primary_key: :domain
+  belongs_to :unavailable_domain, foreign_key: :domain, primary_key: :domain # skipcq: RB-RL1031
```
Comment (ClearlyClaire) [inline app/models/instance.rb:19]: What is the `skipcq` for?

File: app/models/instance.rb [hunk 3]
```diff
@@ -10,10 +10,13 @@
 class Instance < ApplicationRecord
   self.primary_key = :domain
 
+  attr_accessor :failure_days
+
```
Comment (ClearlyClaire) [inline app/models/instance.rb:14]: Finding this attribute yet nothing referencing it in this model is surprising, although I understand it's set in the controller for performance purposes…

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
In addition to the inline comments I made:
- you apparently added an unrelated `bin/deepsource` binary file
- maybe clearer names for `warning_domains`, `warning_domains_map` and `exhausted_deliveries_days` but I don't have better suggestions and honestly, that's nitpicking

File: app/models/instance.rb (hunk 2 above)
Comment (Gargron) [reply app/models/instance.rb:19]: I believe skipcq is for DeepSource, and I think in this line it would complain about lack of inverse association

**Review by Gargron — APPROVED** [review]

File: app/lib/delivery_failure_tracker.rb (hunk 1 above)
Comment (ClearlyClaire) [reply app/lib/delivery_failure_tracker.rb:72]: Still a bit uneasy with this but I guess it is fine.

**Review by ClearlyClaire — APPROVED** [review]

**Comment by noellabo** [comment]
Renamed the class method.

**Review by ClearlyClaire — APPROVED** [review]

**Pull request merged**